### PR TITLE
Introducing Env

### PR DESCRIPTION
**Pull request by guimarqu**
- mandatory argument for the algorithms 
- remove all global mutable variables (!)

In the future, `Env` will allows us to simplify algorithm architecture by moving storage in it

### CODE REVIEW

File: src/decomposition.jl [hunk 1]
```diff
@@ -419,8 +432,8 @@ function getoptbuilder(prob::Problem, ann::BD.Annotation)
 end
 
 function buildformulations!(
-    prob::Problem, annotations::Annotations, reform::Reformulation, parent,
-    node::BD.Root
+    prob::Problem, reform::Reformulation, env::Env, annotations::Annotations, parent,
+    node::BD.Root,
```
Comment (vitornesello) [inline src/decomposition.jl:436]: extra comma

File: src/MathProg/MOIinterface.jl [hunk 2]
```diff
@@ -231,8 +231,8 @@ function get_primal_solutions(form::F, optimizer::MoiOptimizer) where {F <: Form
             kind = _getcolunakind(moirec)
             val = MOI.get(inner, MOI.VariablePrimal(res_idx), moi_index)
             solcost += val * getcurcost(form, id)
-            val = round(val, digits = Coluna._params_.tol_digits)
-            if abs(val) > Coluna._params_.tol
+            val = round(val, digits = Coluna.TOL_DIGITS)
+            if abs(val) > Coluna.TOL
```
[the remark this answers is not in the document]
Comment (guimarqu) [reply src/MathProg/MOIinterface.jl:235]: It's defined in Coluna module (not MathProg)

**Comment by rrsadykov** [comment]
I do not get the point of introducing environment. Is it just to store parameters? Storages are not global, they "belong" to a model, i.e. to the master, to a subproblem. 

We need to create a document explaining the architecture of Coluna so that the changes to architecture are clear. Where is a place for such a document?

For the moment, we have models and algorithms. A model contain user data (formulation, etc) and computed data (storage). Algorithms are immutable and contain just the parameters, and they know child algorithms. I think it is good to have parameters local to algorithms. A user can pass either global parameters, which then are "localized" before solving (high-level usage). A user can also "fine-tune" local parameters of every algorithm, as it is done now (low-level usage). What is the advantage of having global parameters in an environment?

**Comment by guimarqu** [comment]
> We need to create a document explaining the architecture of Coluna so that the changes to architecture are clear. Where is a place for such a document?

I'm writing the document, I'll send you it when I finish the first version. 
I don't plan to change storages before.

> What is the advantage of having global parameters in an environment?

The main goal is to remove all global **mutable** variables such as `_params_` or `_globals_`.

**Review by vitornesello — APPROVED** [review]

**Pull request merged**